lib/level/posts/post.ex

```diff
@@ -0,0 +1,35 @@
+defmodule Level.Posts.Post do
+  @moduledoc """
+  The Post schema.
+  """
+
+  use Ecto.Schema
+
+  import Ecto.Changeset
+
+  alias Level.Spaces.Space
+  alias Level.Spaces.User
+
+  @type t :: %__MODULE__{}
+  @primary_key {:id, :binary_id, autogenerate: true}
+  @foreign_key_type :binary_id
+
+  schema "posts" do
+    field :state, :string, read_after_writes: true
+    field :body, :string
+
+    belongs_to :space, Space
+    belongs_to :user, User
+
+    timestamps()
+  end
+
+  @doc """
+  Generates a changeset for creating a new post.
+  """
+  def create_changeset(struct, attrs \\ %{}) do
+    struct
+    |> cast(attrs, [:space_id, :user_id, :body])
+    |> validate_required([:body])
```

Comment:
So I'm not 100% familiar with your terminology here, but wouldn't it stand to reason that you really want a `user_id` and a `space_id` required along with your post? Otherwise, you might get orphaned records that can't find their associated user or space, right?

Reply:
Great question. I generally only add these types of validations on user-supplied attributes. In this case, the user does not have control over whether `user_id` and `space_id` get set -- the logic that sets those foreign keys in encapsulated in the implementation of `create_post`.

If a bug were to be introduced that stopped setting one of these foreign keys, then I would rather an exception get thrown than a `User ID is required` validation error get propagated back to the user (since the user has no power to manipulate that field on their own). The `NOT NULL` database constraints look after referential integrity in this case and would cause an exception.

Reply:
Oh, yeah, that makes total sense. Since these keys were all part of `attrs` I assumed they were all coming in together as part of a POST request, and therefore things might be able to get weird somehow. And that's also a really good point about what errors are shown to users.

Thanks for such a thorough reply!! 👍